src/server/scripts/Outland/zone_nagrand.cpp

```diff
@@ -636,6 +635,292 @@ class go_warmaul_prison : public GameObjectScript
+    NPC_KIL_SORROW_CULTIST           = 17147,
+    NPC_KIL_SORROW_DEATHSWORN        = 17148,
+    NPC_GISELDA_THE_CRONE            = 18391,
+    NPC_WARMAUL_REAVER               = 17138,
+    NPC_WARMAUL_SHAMAN               = 18064
+};
+
+class npc_nagrand_banner : public CreatureScript
+{
+public:
+    npc_nagrand_banner() : CreatureScript("npc_nagrand_banner") { }
+
+    struct npc_nagrand_bannerAI : public ScriptedAI
+    {
+        npc_nagrand_bannerAI(Creature* creature) : ScriptedAI(creature), bannered(false)
+        {
+            scheduler.SetValidator([this]
+            {
+                return !me->HasUnitState(UNIT_STATE_CASTING);
+            });
+        }
+
+        void Reset() override
+        {
+            scheduler.CancelAll();
+        }
+
+        void SpellHit(Unit* /*caster*/, SpellInfo const* spellInfo) override
+        {
+            if (spellInfo->Id == SPELL_PLANT_WARMAUL_OGRE_BANNER || spellInfo->Id == SPELL_PLANT_KIL_SORROW_BANNER)
+                bannered = true;
+        }
+
+        void UpdateAI(uint32 diff) override
+        {
+            if (!UpdateVictim())
+                return;
+
+            scheduler.Update(diff, [this]
+            {
+                DoMeleeAttackIfReady();
+            });
+        }
+
+        bool IsBannered()
+        {
+            return bannered;
+        }
+
+    protected:
+        TaskScheduler scheduler;
+
+    private:
+        bool bannered;
+    };
+
+    struct npc_kil_sorrow_spellbinderAI : public npc_nagrand_bannerAI
+    {
+        enum KilSorrowSpellBinder
+        {
+            SPELL_ARCANE_MISSILES = 34447,
+            SPELL_CHAINS_OF_ICE   = 22744,
+            SPELL_COUNTERSPELL    = 31999
+        };
+
+        npc_kil_sorrow_spellbinderAI(Creature* creature) : npc_nagrand_bannerAI(creature), has_fled(false), interrupt_cooldown(20000)
+        {
+        }
+
+        void JustEngagedWith(Unit* /*who*/) override
+        {
+            has_fled = false;
+            interrupt_cooldown = 20000;
+            scheduler
+                .Schedule(Seconds(0), [this](TaskContext ArcaneMissiles)
+                {
+                    DoCastVictim(SPELL_ARCANE_MISSILES);
+                    ArcaneMissiles.Repeat(Milliseconds(2400), Milliseconds(3800));
+                })
+                .Schedule(Seconds(3), Seconds(6), [this](TaskContext ChainsOfIce)
+                {
+                    if (Unit* target = SelectTarget(SELECT_TARGET_RANDOM))
+                        DoCast(target, SPELL_CHAINS_OF_ICE, true);
+                    ChainsOfIce.Repeat(Seconds(20),Seconds(25));
+                });
+        }
+
+        void UpdateAI(uint32 diff) override
+        {
+            if (!UpdateVictim())
+                return;
+
+            interrupt_cooldown += diff;
+            if (me->HasUnitState(UNIT_STATE_CASTING))
+                return;
+
+            if (me->GetVictim()->HasUnitState(UNIT_STATE_CASTING) && interrupt_cooldown > 25000)
```

Comment:
Note: this currently crashes because UpdateVictim() returns true even if me->GetVictim() returns nullptr, see https://github.com/TrinityCore/TrinityCore/pull/22875 for a proposed change to change that behavior

Reply:
Adding a NULL check in https://github.com/TrinityCore/TrinityCore/pull/22878